network/src/sync/memory_pool.rs

```diff
@@ -68,93 +66,69 @@ impl<S: Storage + Send + core::marker::Sync + 'static> Node<S> {
         &self,
         source: SocketAddr,
         transaction: Vec<u8>,
-    ) -> Result<(), NetworkError> {
+    ) -> Result<()> {
         if let Ok(tx) = Tx::read(&*transaction) {
-            let insertion = {
-                let storage = self.expect_sync().storage();
-
-                if !self.expect_sync().consensus.verify_transaction(&tx)? {
-                    error!("Received a transaction that was invalid");
-                    return Ok(());
-                }
-
-                if tx.value_balance.is_negative() {
-                    error!("Received a transaction that was a coinbase transaction");
-                    return Ok(());
-                }
-
-                let entry = Entry::<Tx> {
-                    size_in_bytes: transaction.len(),
-                    transaction: tx,
-                };
-
-                self.expect_sync().memory_pool().insert(storage, entry).await
-            };
-
-            if let Ok(inserted) = insertion {
-                if inserted.is_some() {
-                    info!("Transaction added to memory pool.");
-                    self.propagate_memory_pool_transaction(transaction, source).await;
-                }
+            let inserted = self.expect_sync().consensus.receive_transaction(tx.serialize()?).await;
+
+            if inserted {
+                info!("Transaction added to memory pool.");
+                self.propagate_memory_pool_transaction(transaction, source).await;
             }
         }
 
         Ok(())
     }
 
     /// A peer has requested our memory pool transactions.
-    pub(crate) async fn received_get_memory_pool(&self, remote_address: SocketAddr) {
-        // TODO (howardwu): This should have been written with Rayon - it is easily parallelizable.
-        let transactions = {
-            let mut txs = vec![];
-
-            for entry in self.expect_sync().memory_pool().transactions.inner().values() {
-                if let Ok(transaction_bytes) = to_bytes![entry.transaction] {
-                    txs.push(transaction_bytes);
-                }
-            }
-
-            txs
-        };
+    pub(crate) async fn received_get_memory_pool(&self, remote_address: SocketAddr) -> Result<()> {
+        let transactions = self
+            .expect_sync()
+            .consensus
+            .fetch_memory_pool()
+            .await?
+            .into_iter()
+            .map(|x| {
+                let mut out = vec![];
+                x.write(&mut out)?;
+                Ok(out)
```

Comment:
What is `x`? What is `out`? Short names like these need to be made more clear, e.g. here it's presumably `tx` and `digest`.

Reply:
single use lambda variables? i take liberty in making them short so you can grok the lambda (ideally in one line). `out` as a return preparation is a common pattern

Reply:
I get the principle and also like short variable names myself - I had no issue realizing what is what, but it might not be that straightforward to future readers.